src/sentry/models/notificationmessage.py

```diff
@@ -0,0 +1,97 @@
+
+    If any notification message is in relation to another notification message, you should be able to find the original
+    notification message through the parent notification message relation.
+    """
+
+    __relocation_scope__ = RelocationScope.Excluded
+
+    # Related information regarding failed notifications.
+    # Leveraged to help give the user visibility into notifications that are consistently failing.
+    error_details = JSONField(null=True)
+    error_code = IntegerField(null=True, db_index=True)
+
+    # Resulting identifier from the vendor that can be leveraged for future interaction with the notification.
+    message_identifier = CharField(null=True)
+    # Reference to another notification if we choose to modify the original message or reply to it (like start a thread)
+    parent_notification_message = FlexibleForeignKey("self", null=True)
+
+    # Related information regarding Alert Rules (Metric Alerts)
+    incident = FlexibleForeignKey("sentry.Incident", null=True)
+    trigger_action = FlexibleForeignKey("sentry.AlertRuleTriggerAction", null=True)
+
+    # Related information regarding Rules (Issue Alerts)
+    rule_fire_history = FlexibleForeignKey("sentry.RuleFireHistory", null=True)
+    rule_action_uuid = CharField(null=True, db_index=True)
+
+    date_added = DateTimeField(default=timezone.now)
+
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_notificationmessage"
+        # A notification message should exist for either issue or metric alert, but never both
+        constraints = [
+            CheckConstraint(
+                check=(
+                    (
+                        Q(incident__isnull=False, trigger_action__isnull=False)
+                        & Q(rule_fire_history__isnull=True, rule_action_uuid__isnull=True)
+                    )
+                    | (
+                        Q(incident__isnull=True, trigger_action__isnull=True)
+                        & Q(rule_fire_history__isnull=False, rule_action_uuid__isnull=False)
+                    )
+                ),
+                name="notification_for_issue_xor_metric_alert",
+            ),
+            UniqueConstraint(
+                fields=("incident", "trigger_action"),
+                condition=Q(
+                    error_code__isnull=True,
+                    parent_notification_message__isnull=True,
+                    incident__isnull=False,
+                    trigger_action__isnull=False,
+                ),
+                name="singular_parent_message_per_incident_and_trigger_action",
+            ),
+            UniqueConstraint(
+                fields=("rule_fire_history", "rule_action_uuid"),
+                condition=Q(
+                    error_code__isnull=True,
+                    parent_notification_message__isnull=True,
+                    rule_fire_history__isnull=False,
+                    rule_action_uuid__isnull=False,
```

Comment:
Nit: Jfyi, `nulls` are always considered unique, so we don't actually have to specify the `isnull=False` here, since any rows that have nulls in this cols will automatically never violate the unique constraint

Reply:
Ohhh got it! Is this more dangerous now for our migration, or it's fine and this is just explicit?

Reply:
It's not any more dangerous. I'm not sure if it means that it's harder for the planner to use the index or anything - like will it only use this index if you explicitly pass a query like that? It might be smart enough to understand that a condition that partially matches can still use the index though